packages/core/src/session.ts

```diff
@@ -508,7 +519,12 @@ class Session {
    * @returns {void}
    */
   _onDatabaseNameResolved (database?: string): void {
+    this._databaseGuess = database
     if (!this._databaseNameResolved) {
+      if (this._homeDatabaseCallback != null) {
+        // eslint-disable-next-line
+        this._homeDatabaseCallback((this._impersonatedUser ? 'basic:' + this._impersonatedUser : undefined)  ?? this._auth?.cacheKey ?? 'DEFAULT', database)
```

Comment:
For better maintainability, this logic should live somewhere central (close to all the other logic that's concerned with home db caching and cache keying).

(There's a copy of this line further down - won't comment again)